Add stub for new jit-less js->wasm entrypoint

justinmichaud opened the pull request:
#### dc794ece05e682e4fe156bdea696620bc3727261
<pre>
Add stub for new jit-less js-&gt;wasm entrypoint
<a href="https://bugs.webkit.org/show_bug.cgi?id=272176">https://bugs.webkit.org/show_bug.cgi?id=272176</a>
<a href="https://rdar.apple.com/125923445">rdar://125923445</a>

Reviewed by Mark Lam and Yusuke Suzuki.

A subsequent patch will fill this out with a new jit-less js-&gt;wasm
stub, but this patch handles all of the mechanical bits.

This is split out because this patch was causing some linker issues,
so this will make it easier to isolate any fallout from landing.

* Source/JavaScriptCore/assembler/CPU.h:
(JSC::isJSValue3264):
* Source/JavaScriptCore/assembler/JITOperationList.cpp:
(JSC::llintOperations):
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/jit/RegisterSet.cpp:
(JSC::RegisterSetBuilder::wasmPinnedRegisters):
* Source/JavaScriptCore/llint/InPlaceInterpreter.h:
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::logWasmPrologue):
* Source/JavaScriptCore/llint/LLIntSlowPaths.h:
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
(JSC::initialize):
* Source/JavaScriptCore/runtime/JSCPtrTag.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::initializeCallees):
* Source/JavaScriptCore/wasm/WasmBBQPlan.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp:
(JSC::Wasm::CalleeGroup::CalleeGroup):
* Source/JavaScriptCore/wasm/WasmCalleeGroup.h:
* Source/JavaScriptCore/wasm/WasmIPIntPlan.cpp:
(JSC::Wasm::IPIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmIPIntPlan.h:
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmLLIntPlan.h:
* Source/bmalloc/bmalloc/Gigacage.h:
(Gigacage::maxSize):
(Gigacage::mask):

Canonical link: <a href="https://commits.webkit.org/277125@main">https://commits.webkit.org/277125@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/40c1cd42a454c1027906ed1292f2617a45cb3ce6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46650 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25810 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49260 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49327 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42695 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30170 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23272 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38035 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47229 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22809 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40200 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19303 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20174 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41331 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4695 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/39897 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42942 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41694 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51181 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/46132 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21657 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18055 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45294 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 19 flakes 84 failures 1 missing results; Uploaded test results; 6 flakes 76 failures 1 missing results; Running compile-webkit-without-change") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22947 "Built successfully") | | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44251 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests (failure); 1 api test failed or timed out; Running compile-webkit-without-change") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/10330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23402 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/53275 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6547 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22650 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10932 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/assembler/CPU.h:
```diff
@@ -124,6 +124,15 @@ constexpr bool isRISCV64()
 #endif
 }
 
+constexpr bool isJSValue3264()
```
What info is this mean to convey?  Why not use `!is64Bit()` instead?

MenloDorian replied:
My first thought is that this should be named `is32Bit()` instead.  But then, I realized you can just do `!is64Bit()`.  Hence, I asked why this is needed.

MenloDorian commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -703,6 +704,11 @@ end
+# extern "C" __attribute__((__used__)) __attribute__((visibility("hidden"))) void cProbeCallbackFunction(uint64_t i);
+# LLIntSlowPaths.cpp:
+# extern "C" void cProbeCallbackFunction(uint64_t i) {}
+#
```
What's the intent of this comment?  The LLInt `probe` macro below is not full featured like the JIT probe.  This is meant to be used as a convenience debugging only.  Unlike with JIT probes, the protocol of the LLInt probe C callback can take any form of arguments you want.  It's up to the probe user.

justinmichaud replied:
The purpose is to make it easy to add a probe without having to figure out why you broke the linker

MenloDorian replied:
So, you're documenting an example here?  Please indent this by 4 spaces to match the probe example above.  I think the visual cue helps provide context.  Thanks.

MenloDorian commented on Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
```diff
@@ -440,6 +440,17 @@ static UGPRPair entryOSR(CodeBlock* codeBlock, const char*, EntryKind)
 }
 #endif // ENABLE(JIT)
 
+extern "C" void logWasmPrologue(uint64_t i, uint64_t* fp, uint64_t* sp)
```
Can you wrap this in `#if LLINT_TRACING`?  It's not needed otherwise.

MenloDorian commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -571,6 +581,24 @@ end
+    if JSVALUE64
+        andq 0xffffffff, r
+    end
+end
```
I'm nervous about defining this now without seeing how it's used.  Depending on the use, we might want to rely on offlineasm defining some semantics on certain instructions to achieve this.  Using a macro may result in less optimal code.

Can you defer this to the next patch?  I don't think this affects linkage issues.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmCallee.h:
```diff
@@ -53,7 +53,7 @@ class LLIntOffsetsExtractor;
 namespace Wasm {
 
 class Callee : public NativeCallee {
-    WTF_MAKE_TZONE_ALLOCATED(Callee);
+    WTF_MAKE_FAST_ALLOCATED_WITH_HEAP_IDENTIFIER(Callee);
```
Why is this needed?  Let's ping Michael on this.

justinmichaud replied:
The new JIT callee will allocate metadata directly at the end of Callee, which I am assuming doesn't fit the goal of TZone

MenloDorian replied:
Please add a comment in the base class JITCallee to indicate the subclass that is variable sized.  That way, someone won't innocently put this back in the TZone.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmCallee.h:
```diff
@@ -89,11 +89,13 @@ class Callee : public NativeCallee {
 
 #if ENABLE(JIT)
 class JITCallee : public Callee {
-    WTF_MAKE_TZONE_ALLOCATED(JITCallee);
+    WTF_MAKE_FAST_ALLOCATED_WITH_HEAP_IDENTIFIER(JITCallee);
```
Ditto.

MenloDorian reviewed (approved):
r=me with fixes.

Constellation commented on Source/JavaScriptCore/assembler/CPU.h:
```diff
@@ -124,6 +124,15 @@ constexpr bool isRISCV64()
 #endif
 }
 
+constexpr bool isJSValue3264()
```
Let's name it `isJSValue32_64()`.

MenloDorian replied:
I think this can be removed.  It's not needed because we can use `is64Bit()`.  Alternately, let's just name this `is32Bit()` to match `is64Bit()`.

Constellation reviewed (approved):
r=me

justinmichaud committed:
Add stub for new jit-less js->wasm entrypoint
https://bugs.webkit.org/show_bug.cgi?id=272176
rdar://125923445

Reviewed by Mark Lam and Yusuke Suzuki.

A subsequent patch will fill this out with a new jit-less js->wasm
stub, but this patch handles all of the mechanical bits.

This is split out because this patch was causing some linker issues,
so this will make it easier to isolate any fallout from landing.

* Source/JavaScriptCore/assembler/CPU.h:
(JSC::isJSValue3264):
* Source/JavaScriptCore/assembler/JITOperationList.cpp:
(JSC::llintOperations):
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/jit/RegisterSet.cpp:
(JSC::RegisterSetBuilder::wasmPinnedRegisters):
* Source/JavaScriptCore/llint/InPlaceInterpreter.h:
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::logWasmPrologue):
* Source/JavaScriptCore/llint/LLIntSlowPaths.h:
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
(JSC::initialize):
* Source/JavaScriptCore/runtime/JSCPtrTag.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::initializeCallees):
* Source/JavaScriptCore/wasm/WasmBBQPlan.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp:
(JSC::Wasm::CalleeGroup::CalleeGroup):
* Source/JavaScriptCore/wasm/WasmCalleeGroup.h:
* Source/JavaScriptCore/wasm/WasmIPIntPlan.cpp:
(JSC::Wasm::IPIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmIPIntPlan.h:
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmLLIntPlan.h:
* Source/bmalloc/bmalloc/Gigacage.h:
(Gigacage::maxSize):
(Gigacage::mask):

Canonical link: https://commits.webkit.org/277125@main

rkirsling commented on Source/JavaScriptCore/wasm/WasmCalleeGroup.h:
```diff
@@ -202,7 +202,7 @@ class CalleeGroup final : public ThreadSafeRefCounted<CalleeGroup> {
     RefPtr<IPIntCallees> m_ipintCallees;
     RefPtr<LLIntCallees> m_llintCallees;
-    HashMap<uint32_t, RefPtr<JSEntrypointCallee>, DefaultHash<uint32_t>, WTF::UnsignedWithZeroKeyHashTraits<uint32_t>> m_jsEntrypointCallees;
+    HashMap<uint32_t, RefPtr<JITCallee>, DefaultHash<uint32_t>, WTF::UnsignedWithZeroKeyHashTraits<uint32_t>> m_jsEntrypointCallees;
```
Hmm, so we have a `!ENABLE(JIT)` stub for JSEntrypointCallee but not JITCallee, so this breaks the `!ENABLE(JIT) && ENABLE(WEBASSEMBLY)` build. `m_jsEntrypointCallees` itself is used outside of JIT paths, so does this mean that JITCallee should no longer be JIT-exclusive either?
